woocommerce/woocommerce
# Add 2 new GitHub Actions workflows to include COT E2E and API automated tests in CI.

### All Submissions:

-   [ x ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ x ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ x ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?


### Changes proposed in this Pull Request:

Two GitHub Actions workflows were created to run E2E and API automated tests in an environment with COT enabled in our pipelines. In concrete, these are the workflows:
- **cot-pr-build-and-e2e-tests.yml:** Runs E2E and API tests in an environment with COT enabled every time a new PR is created with the label `focus: custom order tables`. This way we can find potential issues caused by new changes before they are merged.
- **cot-build-and-e2e-tests-daily.yml:** Runs E2E and API tests daily at 2:30 AM UTC in the `trunk` branch. This way we can continuously test the current COT integration and also find out potential intermittent issues since we would have a good set of test runs that we might use to find patterns in the test outcomes.

Note: These two workflows might be temporary until the COT project is completed. We might want to decide to remove them or refactor and leave them afterwards when the time comes.

### How to test the changes in this Pull Request:

**NOTE:** I'm afraid you won't be able to test these scenarios in this repository because:
- The scheduled workflows only run in the **default** branch (trunk, in our case), as stated in the [official documentation](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule):
> Scheduled workflows run on the latest commit on the default or base branch.

This means that we won't be able to test the daily COT tests until they are merged into trunk.

- For some reason (I'm new to GitHub Actions, so there must be any explanation that I'm missing), seems that workflows are reused when we create 2 PRs from the same base branch to different target branches. This PR is created to be merged into `trunk`, so we might not be able to test properly what happens when we create another PR at the same time against `cot-main`.

I suggest testing the following instructions in a forked repository like this one, where there is already a [PR with the changes](https://github.com/alopezari/woocommerce/pull/6).

Three different tests can be defined for this PR:

**Scenario: New workflow is triggered when creating a PR labeled as 'focus: custom order tables'**
1. Add any change to the repository.
2. Create a Pull Request and label it as `focus: custom order tables`.
4. Check that the new GitHub Actions workflow `Run tests against PR in an environment with COT enabled` is triggered.

**Scenario: New workflow is not triggered when creating a PR without the 'focus: custom order tables' label**
1. Add any change to the repository.
2. Create a Pull Request and make sure to not add the `focus: custom order tables` label.
3. Check that none of the new GitHub Actions workflows - `Run tests against PR in an environment with COT enabled` and `Run daily tests in an environment with COT enabled` - is triggered.

**Scenario: New workflow is triggered every day around 2:30 AM UTC**
1. Go to the [GitHub Actions page](https://github.com/woocommerce/woocommerce/actions)
2. Check the `Run daily tests in an environment with COT enabled` was triggered every day around 2:30 AM UTC. Note: there might be a slight offset in the execution time, so the workflow might not be triggered exactly at 2:30 AM UTC, but it should be close.

### Other information:

-   [ x ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ N/A ] Have you written new tests for your changes, as applicable?
-   [ x ] Have you successfully run tests with your changes locally?
-   [ N/A? ] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

### FOR PR REVIEWER ONLY:
